// osu-sr-calculator/dist/Objects/osu!/HitObjects/DifficultyHitObject.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const HitObject_1 = require("./HitObject");
/**
 * HitObject with additional difficulty properties required to calculate the star rating
 */
class DifficultyHitObject extends HitObject_1.HitObject {
    /**
     *
     * @param currentObject HitObject to create a DifficultyHitObject from
     * @param lastObject HitObject before the current HitObject
     * @param lastLastObject HitObject before the lastObject
     * @param travelDistance The calculated Travel Distance for this HitObject
     * @param jumpDistance The calculated Jump Distance for this HitObject
     * @param angle The calculated Angle for this HitObject
     * @param deltaTime The calculated Deltatime for this HitObject
     * @param strainTime The calculated Straintime for this hitObject
     */
    constructor(currentObject, lastObject, lastLastObject, travelDistance, jumpDistance, angle, deltaTime, strainTime) {
        super(currentObject.Position, currentObject.StartTime, currentObject.Radius);
        this.TravelDistance = travelDistance;
        this.JumpDistance = jumpDistance;
        this.Angle = angle;
        this.DeltaTime = deltaTime;
        this.StrainTime = strainTime;
        this.CurrentObject = currentObject;
        this.LastObject = lastObject;
        this.LastLastObject = lastLastObject;
    }
    ;
}
exports.DifficultyHitObject = DifficultyHitObject;
;
